pytest/README.md

```diff
@@ -0,0 +1,161 @@
+# LISAv3 via pytest-lisa
+
+[Pytest](https://docs.pytest.org/en/stable/) is an [incredibly
+popular](https://docs.pytest.org/en/stable/talks.html) MIT licensed open source
+Python testing framework. It has a thriving community and plugin framework, with
+[over 750 plugins](https://plugincompat.herokuapp.com/). There is even a YAML
+example of writing a Domain Specific Language
+[DSL](https://docs.pytest.org/en/stable/example/nonpython.html#yaml-plugin) for
+specifying tests. Instead of writing yet another test framework, LISAv3 could be
+written as pytest-lisa, a [plugin for
+Pytest](https://docs.pytest.org/en/stable/writing_plugins.html) which implements
+our requirements. In fact, most of Pytest itself is implemented via [built-in
+plugins](https://docs.pytest.org/en/stable/plugins.html), providing us with a
+lot to leverage.
+
+The [fundamental features](https://www.youtube.com/watch?v=CMuSn9cofbI) of
+Pytest match our needs very well:
+
+* Automatic test discovery, no boiler-plate test code
+* Useful information when a test fails (assertions are introspected)
+* Test parameterization
+* Modular setup/teardown via fixtures
+* Customizable (as detailed above)
+
+So all the logic for discovering, running, skipping based on requirements, and
+reporting the tests is already written and maintained by the greater open source
+community, leaving us to focus on the hard and unique problem: creating an API
+to launch the necessary nodes. It would also allow us the space to abstract the
+installation of tools required by tests. In this way, LISAv3 could solve the
+difficulties we have at hand without creating yet another unit test framework.
+
+## Design
+
+### pytest-mark
+
+The [pytest-mark](https://docs.pytest.org/en/stable/mark.html) already provides
+functionality for adding metadata to tests, where we specifically want:
+
+* Owner
+* Category
+* Area
+* Tags
+* Priority
+
+We could simply reuse this built-in plugin with minimal logic to enforce our
+required metadata, with sane defaults (such as setting the area to the name of
+the module), and to list statistics about our test coverage.
+
+It also through pytest-mark that [skipping
+functionality](https://docs.pytest.org/en/stable/skipping.html) exists, which we
+would leverage for ensuring our environmental requirements are met.
+
+Note that Pytest leverages Python’s docstrings for built-in documentation (and
+can even run tests discovered in such strings, like doctest).
+
+### Fixtures
+
+Pytest supports [fixtures](https://docs.pytest.org/en/stable/fixture.html),
```

Comment:
Does it support to control flow to match test cases and environment? The environment may be defined on runtime by configuration.

Reply:
You'll see in my design document (the readme) that setting up environments based on requirements is the _primary_ work that we need to implement, either on top of Pytest or on top of a brand-new testing framework. I believe my demo proved that this is entirely feasible, and since it's the work that we should be focusing on, letting Pytest do the heavy lifting will make our lives much easier.